### ISSUE TITLE
Bug: [no-unused-var] issue with Angular app

### ISSUE THREAD

**Issue by huineng** [typescript-eslint/typescript-eslint#9719]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://github.com/typescript-eslint/typescript-eslint/blob/aba6e37c4050a2031ee4a5340cdbabe25dd5adc4/packages/eslint-plugin/src/util/collectUnusedVariables.ts#L151

### Repro Code

```TypeScript
import { InjectionToken as InjectionToken } from '@angular/core';
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unused-vars": [
        "error",
        {
            "args": "all",
            "argsIgnorePattern": "^_",
            "caughtErrors": "all",
            "caughtErrorsIgnorePattern": "^_",
            "destructuredArrayIgnorePattern": "^_",
            "varsIgnorePattern": "^_",
            "ignoreRestSiblings": true
        }
    ],
  },
};
```

### tsconfig

```jsonc
{
    "compilerOptions": {
        "alwaysStrict": true,
        "baseUrl": ".",
        "emitDecoratorMetadata": true,
        "experimentalDecorators": true,
        "module": "esnext",
        "moduleResolution": "node",
        "noImplicitAny": true,
        "noImplicitReturns": true,
        "noImplicitThis": true,
        "noUnusedLocals": true,
        "noUnusedParameters": true,
        "removeComments": true,
        "resolveJsonModule": true,
        "esModuleInterop": true,
        "sourceMap": true,
        "strictFunctionTypes": true,
        "strictNullChecks": true,
        "strictPropertyInitialization": true,
        "target": "esnext",
        "version": true,
        "types": ["node", "webpack-env"],
        "lib": ["esnext", "dom"]
    }
}
```

### Expected Result

no linting errors

### Actual Result

I have upgraded to the latest eslint, and i have an angular application for which the first line is

```typescript
import { InjectionToken } from '@angular/core';
```

I get the following error

```log
Oops! Something went wrong! :(

ESLint: 9.8.0

TypeError: Cannot use 'in' operator to search for 'type' in undefined
Occurred while linting /src/client/app.config.ts:1
Rule: "@typescript-eslint/no-unused-vars"
    at UnusedVarsVisitor.markVariableAsUsed (/node_modules/@typescript-eslint/eslint-plugin/dist/util/collectUnusedVariables.js:92:22)
    at UnusedVarsVisitor.TSMappedType (/node_modules/@typescript-eslint/eslint-plugin/dist/util/collectUnusedVariables.js:235:14)
    at UnusedVarsVisitor.visit (/node_modules/@typescript-eslint/scope-manager/dist/referencer/VisitorBase.js:59:21)
    at UnusedVarsVisitor.visitChildren (/node_modules/@typescript-eslint/scope-manager/dist/referencer/VisitorBase.js:46:22)
    at UnusedVarsVisitor.visit (/node_modules/@typescript-eslint/scope-manager/dist/referencer/VisitorBase.js:64:14)
    at UnusedVarsVisitor.visitChildren (/node_modules/@typescript-eslint/scope-manager/dist/referencer/VisitorBase.js:41:30)
    at UnusedVarsVisitor.visit (/node_modules/@typescript-eslint/scope-manager/dist/referencer/VisitorBase.js:64:14)
    at UnusedVarsVisitor.collectUnusedVariables (/node_modules/@typescript-eslint/eslint-plugin/dist/util/collectUnusedVariables.js:30:17)
    at collectVariables (/node_modules/@typescript-eslint/eslint-plugin/dist/util/collectUnusedVariables.js:603:30)
    at collectUnusedVariables (/node_modules/@typescript-eslint/eslint-plugin/dist/rules/no-unused-vars.js:269:65)
   ```
   
   

### Additional Info

for safety it's better also to check on undefined

suggestion

```js
 if (
     variableOrIdentifierOrName &&
      typeof variableOrIdentifierOrName !== 'string' &&
      !('type' in variableOrIdentifierOrName)
    ) {
      variableOrIdentifierOrName.eslintUsed = true;
      return;
    }
```

**Comment by JoshuaKGoldberg** [comment]
> Repro Code

A few scattered config files + a snippet of code isn't a full reproduction. We need to be able to _reproduce_ your issue locally, including seeing the versions of all relevant packages. Please post a way for us to do that, such as a link to your project or a minimum sample app reproduction.

In case it's useful: https://antfu.me/posts/why-reproductions-are-required

**Comment by bradzacher** [comment]
Your repro code also [does not crash in the playground.](https://typescript-eslint.io/play/#ts=5.5.2&fileType=.tsx&code=PQKgsAUApgzgNgSwHYBcAEAiAAigngB1gGMAnBfFAWlkVWCQHtKBXJZmKAE0oDcBDEjAwAuNAG1IaTFBIkGJDABpJaAN4qpUjAIDmQ0drhwlGzdpJ6AkjsYkoABT4oUMpCMwA9APomIms0R8zDoAFigAorLy%2Bph8Rr7%2BWoHBYZFygta2Dk4uJG4G3gmJGJywKCTMRCjMdpwAgrJ8uJny2c6u7hiFyn7F-Bk2rY7teZ3dploIg3YASmUAyggARrR67uXMUCoAvpAAuj0gwJCQwMBoODCUU1mQCAC2%2BPLoqmiWSABWUFUIDEgAKgwANZQJBoPgwN6fb4oX4A4GgtDbNAAMzk9zQAHIsHwkDpmHABMAiK1MQBuSBAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

So your listed reproduction isn't actually a reproduction!

**_Help us to help you_**

**Comment by Josh-Cena** [comment]
If it requires a specific version to reproduce, you can always make a cloneable repository.

**Comment by KnisterPeter** [comment]
This is not only true for angular apps. The rule trips over this type of typescript code:

```ts
export type Prettify<T> = { [P in keyof T]: T[P] } & {};
```

**Comment by KnisterPeter** [comment]
@bradzacher My bad that I just wanted to report that I see exactly the same error outside of angular projects as well. But you do you. 😉

**Comment by bradzacher** [comment]
I plugged your code into the [playground and it worked fine and did not crash](https://typescript-eslint.io/play/#ts=5.5.2&fileType=.tsx&code=KYDwDg9gTgLgBDAnmYcAKVgxgSwGaIA8AKgHxwC8cA3nANppw4B2cA1sIhHnMQLoAuXgz5wAvnABkNMQG4AUPIDGEZgGd4ISnACMsoA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tidmUQAmtAG4BDaKgyRE0aB2iRwYAL4hlQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false).

So likely there's more to your report than a single line of code can encapsulate. For these things we ask for an issue to be filed with a complete reproduction so that we can investigate.

**Comment by huineng** [comment]
You can close this now, after some debugging myself, there error message telling me line 1 and thinking it was an angular support was misleading. At the end i fixed by changing a few interfaces
the problem was that eslint was requiring a "key" for TSMappedType, so i changed from

```
const enum ERoutes {
  home = 'home'
}

type IRoutes = {
  [route in ERoutes]: {
    component?: string;
  };
};
  }
  ```
  
  to 
  
  ```
  type IRoutes = {
    [key: string]: {
        component?: string;
    };
};
```        

which is strange,  because that code i replaced previously never had any issue with lint and displays as ok on the playground, yet the compilation doesn't like it

**Comment by mikestopcontinues** [comment]
After upgrading, I am having this exact issue in a file with:

```ts
type AwaitedVals<T extends UnknownObj> = {
  [K in keyof T]: Awaited<T[K]>;
};
```

I can cause the bug by toggling between the following tsconfig options:

```json5
{
  "compilerOptions": {
    // fails
    "module": "preserve",

    // works
    "module": "es2022",
    "moduleResolution": "node",
  },
}
```

**Comment by kirkwaiblinger** [comment]
@mikestopcontinues Do you have a repro for this? We still don't have anything to work off of here.

The snippet you've provided does not crash in the playground ([link](https://typescript-eslint.io/play/#ts=5.5.4&fileType=.tsx&code=C4TwDgpgBAgg7gQwJbAgEwGoIDYGcA8AKlBAB6oB2auUAqhQNYUD2cFA8gEYBWAfFAF4oAbwBQUKAG0A0lCQUoDCCGYAzKIQC6ALliIU6IjM28A3KIC%2BpoA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tidmUQAmtAG4BDaKgyQA7uKaRwYAL4glQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false))

**Comment by huineng** [comment]
ok, but that link is for version 9.3.0 while the issue is with 9.8.0 and for me it happens with compilation
./node_modules/.bin/eslint while i had no error in vscode , so i guess you might need a custom repo for this

**Comment by moroine** [comment]
I have the same error, but with next here is a reproduction https://codesandbox.io/p/devbox/dry-night-dtwskm?workspaceId=a0b17b0e-6999-4100-9745-a9057585d895&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clzsbn82c00063b6ilfpismga%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clzsbn82c00023b6i3hm9yke9%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clzsbn82c00043b6ifoadbmyl%2522%257D%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clzsbn82c00053b6i7ecoxifd%2522%257D%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clzsbn82c00023b6i3hm9yke9%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clzsbn82c00013b6i1ako4v38%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252FREADME.md%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clzsbo9g0000y3b6iz2vcnn43%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A6%252C%2522startColumn%2522%253A15%252C%2522endLineNumber%2522%253A6%252C%2522endColumn%2522%253A15%257D%255D%252C%2522filepath%2522%253A%2522%252Fpackage.json%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clzsbn82c00023b6i3hm9yke9%2522%252C%2522activeTabId%2522%253A%2522clzsbo9g0000y3b6iz2vcnn43%2522%257D%252C%2522clzsbn82c00053b6i7ecoxifd%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clzsbn82c00053b6i7ecoxifd%2522%257D%252C%2522clzsbn82c00043b6ifoadbmyl%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clzsbn82c00033b6ii864ymfa%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522start%2522%257D%255D%252C%2522id%2522%253A%2522clzsbn82c00043b6ifoadbmyl%2522%252C%2522activeTabId%2522%253A%2522clzsbn82c00033b6ii864ymfa%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D

**Comment by bradzacher** [comment]
@moroine 
```
$ yarn list @typescript-eslint/parser
├─ @typescript-eslint/parser@8.1.0
└─ eslint-config-next@13.5.6
   └─ @typescript-eslint/parser@6.21.0
Done in 0.11s.
```

`ui/.eslintrc.json` has `"extends": ["next"]`.
i.e. your using v6 of our parser with v8 of our lint rules.
it's expected that this state is invalid and crashes.

The latest version of `eslint-config-next` (14.2.5) depends on a _slightly_ more modern version of our packages (v7.2.0) -- but it's still miles behind the latest version (8.1.0) -- it's even miles behind the latest v7 version (7.18.0).
For reference we released 7.2.0 **_FIVE_** months ago

---

Given we don't have a valid repro after a week - I'm going to close this as unreproducible

**Comment by moroine** [comment]
@bradzacher I think that's the point, invalid peer dependencies thanks.
